source/material_model/viscoelastic_plastic.cc

```diff
@@ -0,0 +1,796 @@
+          // given positions. The term reaction_terms[i][c] is the
+          // change in compositional field c at point i.
+          for (unsigned int c=0; c<in.composition[i].size(); ++c)
+            out.reaction_terms[i][c] = 0.0;
+
+          if (in.strain_rate.size())
+            {
+
+              // Calculate the square root of the second moment invariant for the deviatoric strain rate tensor.
+              // The first time this function is called (first iteration of first time step)
+              // a specified "reference" strain rate is used as the returned value would
+              // otherwise be zero.
+              const double edot_ii = ( (this->get_timestep_number() == 0 && strain_rate.norm() <= std::numeric_limits<double>::min() )
+                                       ?
+                                       reference_strain_rate
+                                       :
+                                       std::max(std::sqrt(std::fabs(second_invariant(deviator(strain_rate)))),
+                                                minimum_strain_rate) );
+
+              const std::vector<double> viscosities_pre_yield = linear_viscosities;
+
+              // TODO: Add strain-weakening of cohesion and friction
+              const std::vector<double> coh = cohesions;
+              const std::vector<double> phi = angles_internal_friction;
+
+              // Initialize variables
+              std::vector<double> stresses_viscous(volume_fractions.size());
+              std::vector<double> stresses_yield(volume_fractions.size());
+              std::vector<double> viscosities_drucker_prager(volume_fractions.size());
+              std::vector<double> viscosities_yield(volume_fractions.size());
+              std::vector<double> viscosities_viscoelastic(volume_fractions.size());
+
+              // Loop through all compositions
+              for (unsigned int j=0; j < volume_fractions.size(); ++j)
+                {
+
+                  stresses_viscous[j] = 2. * viscosities_pre_yield[j] * edot_ii;
+
+                  stresses_yield[j] = ( (dim==3)
+                                        ?
+                                        ( 6.0 * coh[j] * std::cos(phi[j]) + 6.0 * std::max(pressure,0.0) * std::sin(phi[j]) )
+                                        / ( std::sqrt(3.0) * (3.0 + std::sin(phi[j]) ) )
+                                        :
+                                        coh[j] * std::cos(phi[j]) + std::max(pressure,0.0) * std::sin(phi[j]) );
+
+                  if ( stresses_viscous[j] >= stresses_yield[j]  )
+                    {
+                      viscosities_drucker_prager[j] = stresses_yield[j] / (2.0 * edot_ii);
+                    }
+                  else
+                    {
+                      viscosities_drucker_prager[j] = viscosities_pre_yield[j];
+                    }
+
+                  viscosities_yield[j] = viscosities_drucker_prager[j];
```

Comment:
why is there a viscosities_yield that is exactly the same as viscosities_drucker_prager?  For later implementation of other yield mechanisms?

Reply:
Yes, exactly. The plan is to add in more yield mechanisms like the stress limiter rheology, rate-state friction, etc.